smarts/core/lanepoints.py

```diff
@@ -0,0 +1,564 @@
+
+    @property
+    def pose(self) -> Pose:
+        return Pose.from_center(tuple(self.pos), self.heading)
+
+    def __hash__(self):
+        return hash(
+            (
+                *self.pos,
+                self.heading,
+                self.lane_width,
+                self.speed_limit,
+                self.lane_id,
+                self.lane_index,
+            )
+        )
+
+    def __eq__(self, other):
+        if not isinstance(other, LanePoint):
+            return False
+        return (
+            (self.pos == other.pos).all()
+            and self.heading == other.heading
+            and self.lane_width == other.lane_width
+            and self.speed_limit == other.speed_limit
+            and self.lane_id == other.lane_id
+            and self.lane_index == other.lane_index
+        )
+
+
+class LinkedLanePoint(NamedTuple):
+    lp: LanePoint = None
+    is_shape_lp: bool = False
+    nexts: List[LinkedLanePoint] = []  # list of next immediate LanePoint(s)
+    # it's a list of LanePoints because a path may branch at junctions
+
+    def __hash__(self):
+        ## distinguish between different continuations here too
+        ## so the lru_cache on _lanepoint_paths_starting_at_lanepoint() below
+        ## doesn't return the wrong set of LanePoints.
+        return hash((self.lp, tuple(nlp.lp for nlp in self.nexts)))
+
+
+class LanePoints:
+    def __init__(self, road_network, spacing, debug=True):
+        self.spacing = spacing
+        self._road_network = road_network
+
+        # XXX: for a big map, may not want to cache ALL of the potential LanePoints
+        #      nor waste time here finding all of them.
+        #      Lanepoints might be generated on demand based upon edges and lookahead.
+        shape_lps = self._shape_lanepoints()
+        self._linked_lanepoints = LanePoints._interpolate_shape_lanepoints(
+            shape_lps, spacing
+        )
+
+        self._lanepoints_kd_tree = LanePoints._build_kd_tree(self._linked_lanepoints)
+
+        self._lanepoints_by_lane_id = defaultdict(list)
+        self._lanepoints_by_edge_id = defaultdict(list)
+        for linked_lp in self._linked_lanepoints:
+            lp_edge_id = self._edge(linked_lp.lp).getID()
+            self._lanepoints_by_lane_id[linked_lp.lp.lane_id].append(linked_lp)
+            self._lanepoints_by_edge_id[lp_edge_id].append(linked_lp)
+
+        self._lanepoints_kd_tree_by_lane_id = {
+            lane_id: LanePoints._build_kd_tree(l_lps)
+            for lane_id, l_lps in self._lanepoints_by_lane_id.items()
+        }
+
+        self._lanepoints_kd_tree_by_edge_id = {
+            edge_id: LanePoints._build_kd_tree(l_lps)
+            for edge_id, l_lps in self._lanepoints_by_edge_id.items()
+        }
+
+    @staticmethod
+    def _build_kd_tree(linked_lps):
+        return KDTree(np.array([l_lp.lp.pos for l_lp in linked_lps]), leaf_size=50)
+
+    @staticmethod
+    def _interpolate_shape_lanepoints(shape_lanepoints, spacing):
```

Comment:
Type hints?

Reply:
For this one and a lot of the others in this class, I did not add them because (a) this is pre-existing code that just got refactored here and (b) these are private methods.  I'd be fine with your making a new Issue to go audit all function prototypes in our existing codebase to add type hints to all of them, but that's not what this PR was about and it is frankly too time consuming and distracting to track down each of them on someone else's pre-existing code when I'm already trying to make a fairly large change.  

As a compromise though, for this review, I'll add some more type hints where I know for sure what they are.